File: Programmazione_1/TDE_Prog1/Codici_esame_orale/Esercizio_orale_Sara.c
#include <stdio.h>
#include <stdlib.h>

#define DIM 100

void CreaMat(int mat1[DIM][DIM], int mat2[DIM][DIM], int dim){
    for(int i=0; i<dim; i++){
        for (int j=0; j<dim; j++){
            mat1[i][j]=(i+1)*(j+1);
            mat2[dim-i-1][dim-1-j]=(i+1)*(j+1);
        }
    }
}

int SommaRiga(int riga1[DIM], int riga2[DIM], int riga[DIM], int j, int dim){
    if (j==dim){
        return 0;
    }

    riga[j]= riga1[j]+riga2[j];
    SommaRiga(riga1, riga2, riga, j+1, dim);
}

int SommaMat(int mat1[DIM][DIM], int mat2[DIM][DIM], int mats[DIM][DIM], int i, int dim){
    int riga[DIM];

    if (i==dim){
        return 0;
    }

    SommaRiga(mat1[i], mat2[i], mats[i], 0, dim);
    SommaMat(mat1, mat2, mats, i+1, dim);
}

int main(){
    int matrice1[DIM][DIM], matrice2[DIM][DIM], dim, matsomma[DIM][DIM];

    printf("Inserire la dimensione della matrice (tra 2 e 100): ");
    scanf("%d", &dim);

    while(dim<2 || dim > 100) {
        printf("Reinserire: ");
        scanf("%d", &dim);
    }

    CreaMat(matrice1, matrice2, dim);
    SommaMat(matrice1, matrice2, matsomma, 0, dim);

    for(int i=0; i<dim; i++){
        for (int j=0; j<dim; j++){
            printf("%d\t", matsomma[i][j]);
        }
        printf("\n");
    }

    return 0;
}

File: Programmazione_1/TDE_Prog1/Codici_esame_orale/Esercizio_orale_Sara.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type MATRICE [][]int

func Sommamat(mat1, mat2 MATRICE) MATRICE {
	var mats MATRICE

	if len(mat1) == 0 {
		return mat2
	} else if len(mat2) == 0 {
		return mat1
	}

	mats = append(mats, SommaElementi(mat1[0], mat2[0]))
	mats = append(mats, Sommamat(mat1[1:], mat2[1:])...)
	return mats
}

func SommaElementi(riga1, riga2 []int) []int {
	var riga []int

	if len(riga1) == 0 || len(riga2) == 0 {
		return riga
	}

	riga = append(riga, riga1[0]+riga2[0])
	riga = append(riga, SommaElementi(riga1[1:], riga2[1:])...)
	return riga
}

func AssegnaMat1() (mat MATRICE) {
	scanner := bufio.NewScanner(os.Stdin)

	fmt.Println("Inserire una riga alla volta la prima matrice, ogni elemento separato da uno spazio:")
	for scanner.Scan() {
		var elementi []int
		riga := scanner.Text()
		parts := strings.Split(riga, " ")

		for _, elemento := range parts {
			n, _ := strconv.Atoi(elemento)
			elementi = append(elementi, n)
		}
		mat = append(mat, elementi)
	}

	return
}

func AssegnaMat2() (mat MATRICE) {
	scanner := bufio.NewScanner(os.Stdin)

	fmt.Println("Inserire una riga alla volta la seconda matrice, ogni elemento separato da uno spazio:")
	for scanner.Scan() {
		var elementi []int
		riga := scanner.Text()
		parts := strings.Split(riga, " ")

		for _, elemento := range parts {
			n, _ := strconv.Atoi(elemento)
			elementi = append(elementi, n)
		}
		mat = append(mat, elementi)
	}

	return
}

func main() {
	var mat1, mat2 MATRICE

	mat1 = AssegnaMat1()
	mat2 = AssegnaMat2()

	mats := Sommamat(mat1, mat2)

	for _, righe := range mats {
		fmt.Println(righe)
	}
}
